### ISSUE TITLE
Social login does not show Terms of Use

### ISSUE THREAD

**Issue by jochenklar** [rdmorganiser/rdmo#161]
An initial registration by passes the terms of use text an dcheckbox. Maybe `SOCIALACCOUNT_AUTO_SIGNUP = False` can help.

**Comment by jmfrenzel** [comment]
Can't reproduce this. We have TOU popup + confirmation running with ORCID as the sole auth method shown to the users. 
`local.py`
```
ACCOUNT_TERMS_OF_USE = True und
ACCOUNT = True
SOCIALACCOUNT = True
SOCIALACCOUNT_SIGNUP = True
SOCIALACCOUNT_AUTO_SIGNUP = True 
```

**Comment by jochenklar** [comment]
Hi @jmfrenzel , so if people have a public email, the TOU is skipped, right? Is that a problem, or can e close here?

**Comment by jmfrenzel** [comment]
That is right. It is a problem since our use must agree to the TOU. 
Solution would be 
1. check if SOCIALACCOUNT_AUTO_SIGNUP = 'True'   
   and   
    ACCOUNT_TERMS_OF_USE = 'True'
1. If  both are 'True' force SOCIALACCOUNT_AUTO_SIGNUP = 'False'

**Comment by MyPyDavid** [comment]
I need to implement this for our instance as well. So far Ive set `SOCIALACCOUNT_AUTO_SIGNUP=True` , think I also need to check for current users to agree to the TOU at login.
Was there something implemented from this issue?

**Comment by jochenklar** [comment]
Try it with `SOCIALACCOUNT_AUTO_SIGNUP=False`? Then it should show the signup form nevertheless. (I hope.)

**Comment by MyPyDavid** [comment]
The auto signup can be interrupted in the adapter, like so for example(?):

    def is_auto_signup_allowed(self, request, sociallogin):
        # If email is specified, check for duplicate and if so, no auto signup.
        auto_signup = app_settings.AUTO_SIGNUP
        if settings.ACCOUNT_TERMS_OF_USE:
            if sociallogin.user.id is None:
                return False
            return ConsentFieldValue.objects.filter(user__id=sociallogin.user.id).exists()
        return auto_signup

**Comment by MyPyDavid** [comment]
I've found a way for showing the ToU at signup and for existing users (after login, which would require some changes to the rdmo `home` view).

The auto signup can be interrupted, then the `SocialSignupForm` can be adapted to (similar to the SignupForm in  [rdmo.accounts.adapter](https://github.com/rdmorganiser/rdmo/blob/1eadee51947399601d85e476b261391826e0f95a/rdmo/accounts/adapter.py#L55)). This `SocialSignupForm` can be used via the setting `SOCIALACCOUNT_FORMS `.
```py
    def is_auto_signup_allowed(self, request, sociallogin):
        # If email is specified, check for duplicate and if so, no auto signup.
        auto_signup = app_settings.AUTO_SIGNUP
        if settings.ACCOUNT_TERMS_OF_USE:
                return False
        return auto_signup
```
```py
    if ACCOUNT_TERMS_OF_USE:
        SOCIALACCOUNT_FORMS = {"signup": "accounts.forms.SocialSignupForm"} # path in rdmo-app
````

For existing users, the use case might be that, when the text of the ToU changes, that users need to re-confirm their consent.
This would require an update consent page (view, form and url)  and a check at login if the ToU is enable and if the user has the `ConsentFieldValue`. I've hacked this now into the rdmo-app but think it could better be in source as well, maybe already via a viewset.

**Comment by jochenklar** [comment]
Yes, nice, but you could just set `AUTO_SIGNUP` to false right?

I discussed with @m6121 something similar to the "update" use case above: LDAP. I think we need to have a signal handler which checks if a user has consented yet and does not let them log in before clicking on a to-be-added update consent page.

In https://github.com/django-daiquiri/daiquiri I used something similar (but only with allauth)
https://github.com/django-daiquiri/daiquiri/blob/master/daiquiri/auth/adapter.py#L28. `account_pending` is just a static landing page in this case.

Maybe this needs to be done with a custom Middleware. We could also disable users on login which have not consented yet, but this might cause trouble with other reasons to disable a user.

**Comment by MyPyDavid** [comment]
Yeah, it works with just `SOCIALACCOUNT_AUTO_SIGNUP=False` and without overwriting the auto_signup_allowed method as well.

**Comment by jochenklar** [comment]
Basically this line: https://github.com/cyface/django-termsandconditions/blob/main/termsandconditions/middleware.py#L37 but with some performance tweaking: set a flag in the session the first time and then check this flag fist. 

In RDMO every path would be protected.

**Comment by jochenklar** [comment]
The middleware will be part of RDMO 2.3.0.
